pydantic/_internal/_decorators.py

```diff
@@ -341,68 +349,83 @@ def gather_decorator_functions(cls: type[Any]) -> DecoratorInfos:
+            partialmethod,
+        )
+
+    while isinstance(func, all):
+        if unwrap_class_static_method and isinstance(func, (classmethod, staticmethod)):
+            func = func.__func__
+        elif isinstance(func, (partial, partialmethod)):
+            func = func.func
+
+    return func
+
+
+def get_function_ref(func: Any) -> str:
+    func = unwrap_wrapped_function(func)
+    return (
+        getattr(func, '__module__', '<No __module__>')
+        + '.'
+        + getattr(func, '__qualname__', f'<No __qualname__: id:{id(func)}>')
+    )
 
 
-def is_classmethod_from_sig(function: Callable[..., Any] | classmethod[Any] | staticmethod[Any]) -> bool:
-    sig = signature(unwrap_unbound_methods(function))
+def is_classmethod_from_sig(function: AnyDecoratorCallable) -> bool:
+    sig = signature(unwrap_wrapped_function(function))
     first = next(iter(sig.parameters.values()), None)
     if first and first.name == 'cls':
         return True
     return False
 
 
-def is_instance_method_from_sig(function: Callable[..., Any] | classmethod[Any] | staticmethod[Any]) -> bool:
-    sig = signature(unwrap_unbound_methods(function))
+def is_instance_method_from_sig(function: AnyDecoratorCallable) -> bool:
+    sig = signature(unwrap_wrapped_function(function))
     first = next(iter(sig.parameters.values()), None)
     if first and first.name == 'self':
         return True
     return False
 
 
 def ensure_classmethod_based_on_signature(
-    function: Callable[..., Any] | classmethod[Any] | staticmethod[Any],
-) -> classmethod[Any] | staticmethod[Any] | Callable[..., Any]:
-    if not isinstance(function, classmethod) and is_classmethod_from_sig(function):
-        return classmethod(function)
+    function: AnyDecoratorCallable,
+) -> Any:
+    if not isinstance(
+        unwrap_wrapped_function(function, unwrap_class_static_method=False), classmethod
+    ) and is_classmethod_from_sig(function):
+        return classmethod(function)  # type: ignore[arg-type]
     return function
 
 
-def check_for_duplicate_validator(
-    function: Callable[..., Any] | classmethod[Any] | staticmethod[Any], allow_reuse: bool
+def check_for_duplicate_decorator_function(
+    function: AnyDecoratorCallable, allow_reuse: bool, type: Literal['validator', 'serialzier']
 ) -> None:
     """
     Warn about validators with duplicated names since without this, they can be overwritten silently
     which generally isn't the intended behaviour, don't run in ipython (see #312) or if `allow_reuse` is True.
     """
     if not allow_reuse and not in_ipython():
-        function = unwrap_unbound_methods(function)
-        ref = f'{function.__module__}::{function.__qualname__}'
+        ref = get_function_ref(function)
         if ref in _FUNCS:
-            warnings.warn(f'duplicate validator function "{ref}"; if this is intended, set `allow_reuse=True`')
+            warnings.warn(f'duplicate {type} function "{ref}"; if this is intended, set `allow_reuse=True`')
```

Comment:
Remember we want to replace this with #4436.

But happy to defer that to a new PR. @hramezani, probably don't start work on #4436 until this is merged.